pkg/gather/gather.go

```diff
@@ -0,0 +1,247 @@
+package gather
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sort"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/insights-operator/pkg/anonymization"
+	"github.com/openshift/insights-operator/pkg/config/configobserver"
+	"github.com/openshift/insights-operator/pkg/gather/clusterconfig"
+	"github.com/openshift/insights-operator/pkg/gather/workloads"
+	"github.com/openshift/insights-operator/pkg/record"
+	"github.com/openshift/insights-operator/pkg/recorder"
+	"github.com/openshift/insights-operator/pkg/utils"
+)
+
+const (
+	archiveMetadataFilename = "insights-operator/gathers"
+	allGatherersConst       = "ALL"
+)
+
+var (
+	programStartTime        time.Time
+	workloadsGathererPeriod = time.Hour * 12
+)
+
+func init() {
+	programStartTime = time.Now()
+}
+
+// GathererFunctionReport contains the information about a specific gathering function
+type GathererFunctionReport struct {
+	FuncName     string      `json:"name"`
+	Duration     int64       `json:"duration_in_ms"`
+	RecordsCount int         `json:"records_count"`
+	Errors       []string    `json:"errors"`
+	Panic        interface{} `json:"panic"`
+}
+
+// ArchiveMetadata contains the information about the archive and all its' gatherers
+type ArchiveMetadata struct {
+	// info about gathering functions
+	StatusReports []GathererFunctionReport `json:"status_reports"`
+	// MemoryAlloc is the amount of memory taken by heap objects after processing the records
+	MemoryAlloc uint64 `json:"memory_alloc_bytes"`
+	// Uptime is the number of seconds from the program start till the point when metadata was created
+	Uptime float64 `json:"uptime_seconds"`
+	// IsGlobalObfuscationEnabled shows if obfuscation(hiding IPs and cluster domain) is enabled
+	IsGlobalObfuscationEnabled bool `json:"is_global_obfuscation_enabled"`
+}
+
+// CreateAllGatherers creates all the gatherers important gatherers and optional ones with their periods
+func CreateAllGatherers(
+	gatherKubeConfig, gatherProtoKubeConfig, metricsGatherKubeConfig *rest.Config,
+	anonymizer *anonymization.Anonymizer,
+) ([]Interface, map[time.Duration]Interface) {
```

Comment:
I find these return values a little bit confusing TBH. I think returning just `[]Interface` (maybe also change the name of the `Interface`?) would be more clear. 

Just Friday's idea - What about adding the last processed time (and maybe the `period` as well?) to each gatherer and access it via some `GetLastProcessedTime` (via the `Interface`)?

Reply:
> I find these return values a little bit confusing TBH. I think returning just []Interface (maybe also change the name of the Interface?) would be more clear.

Well, they are 2 different types of gatherers, first list is processed always, every interval, second - only when it's there time.

> Just Friday's idea - What about adding the last processed time (and maybe the period as well?) to each gatherer and access it via some GetLastProcessedTime (via the Interface)?

Hm, yeah, great idea! The only thing is that we want to make sure that with the new approach some gatherers(now it's only cluster config gatherer) are processed every interval and get to every archive as it is now.

Reply:
Yes I know that you wanted to distinguish the types that's why I suggested the idea. I guess I am saying let's distinguish them on the impl level of the `Interface`. 
And yes the `clusterconfig` one will be kinda special from this point of view.

Reply:
Done